classes/question_ui_renderer.php

```diff
@@ -355,27 +368,62 @@ private function resolve_placeholders(): void {
+                $this->append_html_fragment($element, $rawvalue, LIBXML_NOERROR);
             } else {
-                $rawvalue = $this->placeholders[$key];
-                if (strcasecmp($cleanoption, 'clean') == 0) {
-                    // Allow (X)HTML, but clean using Moodle's clean_text to prevent XSS.
-                    $element = $this->xpath->document->createDocumentFragment();
-                    $element->appendXML(clean_text($rawvalue));
-                } else if (strcasecmp($cleanoption, 'noclean') == 0) {
-                    $element = $this->xpath->document->createDocumentFragment();
-                    $element->appendXML($rawvalue);
-                } else {
-                    if (strcasecmp($cleanoption, 'plain') != 0) {
-                        debugging("Unrecognized placeholder cleaning option: '$cleanoption', using 'plain'");
-                    }
-                    // Treat the value as plain text and don't allow any kind of markup.
-                    // Since we're adding a text node, the DOM handles escaping for us.
-                    $element = new DOMText($rawvalue);
+                if (strcasecmp($cleanoption, 'plain') != 0) {
+                    debugging("Unrecognized placeholder cleaning option: '$cleanoption', using 'plain'");
                 }
-                $pi->parentNode->replaceChild($element, $pi);
+                // Treat the value as plain text and don't allow any kind of markup.
+                // Since we're adding a text node, the DOM handles escaping for us.
+                $element = new DOMText($rawvalue);
+            }
+            $pi->parentNode->replaceChild($element, $pi);
+        }
+    }
+
+    /**
+     * Parses and appends some HTML source to the given fragment.
+     *
+     * @param DOMDocumentFragment $fragment
+     * @param string $html
+     * @param int $options
+     * @return bool
+     * @see DOMDocumentFragment::appendXML() the XML equivalent is provided by PHP, but not HTML :(
+     */
+    private function append_html_fragment(DOMDocumentFragment $fragment, string $html, int $options = 0): bool {
+        $newdoc = new DOMDocument();
+        // Libxml will add html and/or body elements and a DTD declaration without these options.
+        $options |= LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD;
+        // Despite LIBXML_HTML_NOIMPLIED, libxml will wrap a <p>-tag around the html if it doesn't have a root element.
+        if (!$newdoc->loadHTML('<body>' . $html . '</body>', $options)) {
+            return false;
+        }
+
+        /** @var DOMNode $childnode */
+        foreach ($newdoc->documentElement->childNodes as $childnode) {
+            $imported = $fragment->ownerDocument->importNode($childnode, deep: true);
+            if ($imported === false) {
+                debugging('Could not import HTML node from placeholder value');
+                return false;
             }
+            $fragment->appendChild($imported);
         }
```

Comment:
Wenn diese Schleife zwei Durchläufe macht und erst bei dem zweiten Durchlauf mit `return false` abbricht, dann wird der erste Node trotzdem importiert und angehängt, oder nicht? Soll das so sein?

Reply:
Es ist leider nicht dokumentiert, wann `importNode` genau false ausgibt. Ich habe aus `append_html_fragment` jetzt `html_to_fragment` gemacht, und verwerfe das ganze Fragment, wenn `importNode` fehlschlägt.

Insgesamt bedarf das Error-Handling glaube ich noch einer genaueren Betrachtung. PHP's DOM bzw. libxml ist sehr undurchsichtig, was Warnungen, Errors (im Sinne von `trigger_error`), Exceptions und einen `false`-Rückgabewert verursacht. Das dann aber separat.

Reply:
Du könntest die `$imported`-Fragmente auch in einer Liste sammeln und in einer zweiten Schleife anhängen, wenn für keines der Fragmente `false` zurückgegeben wurde. Aber wenn du meinst, dass das so auch sauber ist, dann auch gerne so.

Reply:
Ja, schien mir einfacher/sauberer. 

Ich habe mich übrigens etwas näher mit dem Error-Handling in PHP's DOM und libxml befasst, bin aber zu dem Schluss gekommen, dass es vermutlich schon so passt, wie es gerade ist. Um es besser zu machen müsste man denke ich mit `libxml_use_internal_errors` arbeiten, aber das wirkt sich dann auch wieder auf `clean_text` aus...
[Hier 1 Branch](https://github.com/questionpy-org/moodle-qtype_questionpy/compare/question_ui_renderer-order...renderer-error-handling), aber ich denke der Aufwand lohnt nicht.